Простые перечисления

dmpas opened the pull request:
Создам запрос, чтобы обсуждение в нём вести. Вливать пока не надо.

dmpas committed:
Упрощено описание перечислений.

dmpas committed:
Настройки форматирования по CODESTYLE

dmpas committed:
СтатусСообщения переведён на простое перечисление.

dmpas committed:
Убрана лишняя аннотация при создании перечислений.

Глобальное свойство теперь создаётся по-умолчанию и с тем же именем/псевдонимом
, что и тип перечисления.

EvilBeaver commented on src/ScriptEngine.HostedScript/Library/MessageStatusEnum.cs:
```diff
@@ -8,7 +8,6 @@
 namespace ScriptEngine.HostedScript.Library
 {
     [EnumerationType("СтатусСообщения", "MessageStatus")]
-	[EnumerationGlobalProperty("СтатусСообщения", "MessageStatus")]
```
Все-таки решил убрать?

dmpas replied:
ага, порыскал по синтаксис-помощнику и таки решил убрать :)

EvilBeaver replied:
Вроде как было вложенное перечисление "СвойстваОбъектовМетаданных". Но в любом случае, это более редкое явление и его можно поддержать иным способом, чем принудительно в каждом enum писать лишний атрибут.

dmpas replied:
Там третьим параметром можно указать `false` и свойство не будет создаваться автоматом.

EvilBeaver commented on src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs:
```diff
@@ -37,7 +37,13 @@ public static object ConvertParam(IValue value, Type type)
             }
-            else if (type == typeof(IValue))
+
+            if (Nullable.GetUnderlyingType(type) != null)
```
@dmpas расскажи, пожалуйста, что этот код делает? Что такое Nullable?

dmpas replied:
Nullable - это когда `int?` или `enum?`

то есть, это тип не `enum`, и не `int`, и не `IValue` и ни под одно условие ниже не подпадает.

EvilBeaver replied:
А в каких случаях это условие будет истинным? (Nullable.GetUnderlyingType(type) != null)? Если передали typeof(someEnum)? А почему тогда Nullable? Это не придирки к коду, это я для саморазвития.

dmpas replied:
когда я только начинал, я сделал

``` c#
void SortByValue(SortDirectionEnum? direction = null)
```

оно понятно, что сейчас переделал на правильное 

``` c#
void SortByValue(SortDirectionEnum direction = SortDirectionEnum.Asc)
```

но тогда было так и маршаллер не срабатывал, потому что хотел преобразование в `SortDirectionEnum?` (`Nullable<SortDirectionEnum>`), а не в `SortDirectionEnum`. Просто-напросто не находил такой тип, потому что это другой тип.

Это в параметрах. Сейчас буду отрабатывать ещё возвращаемые значения - там тоже может встретиться что-нибудь вроде `int?`, чтобы вернуть число или Неопределено - там тоже надо будет `Nullable` отработать.

dmpas replied:
То есть, передали мы не `typeof(someEnum)`, а `typeof(someEnum?)`, тогда это условие будет истинным.

dmpas committed:
Переименована аннотация ContextFieldAttribue.

`FieldContextAttribute` переименована в `ContextFieldAttribute` во имя
единообразия.

dmpas committed:
Добавлен тест на поведение перечислений.

dmpas committed:
Настройки форматирования

EvilBeaver commented on src/ScriptEngine.HostedScript/Library/SortDirectionEnum.cs:
```diff
@@ -10,10 +10,10 @@ namespace ScriptEngine.HostedScript.Library
     public enum SortDirectionEnum
     {
-        [FieldContext("Возр", "ASC")]
+        [ContextField("Возр", "ASC")]
```
А может быть, EnumItem?

dmpas replied:
@EvilBeaver `Context`? :) да, слово `Enum` было бы разумно увидеть. С другой стороны, ежу понятно, что это `Enum` и здесь его `Item`. Не знаю.

EvilBeaver replied:
Возможно, просто слово Field у меня не ассоциируется с элементом Enum-а

dmpas committed:
`ContextField` переименован в `EnumItem`.

dmpas commented:
Ну, я вроде всё. Если нет более замечаний, то можно это слить, отдельной допилкой перевести все остальные перечисления из стандартной библиотеки и объявить старую аннотацию устаревшей.

EvilBeaver commented:
@dmpas а сейчас работают оба варианта поиска и включения перечислений, по обеим аннотациям?

dmpas commented:
@EvilBeaver да, я поменял лишь СтатусСообщения и НаправлениеСортировки. По старому варианту ещё куча всякого работает: ХешФункция, ТипУзлаXml, УровеньСжатияZIP и многое другое. К тому же, старый вариант используется для всяких ЦветаКонсоли и Символы, что как бы неверно, ибо это не перечисления, а наборы значений.

EvilBeaver commented:
@dmpas Compilation errors

MainWindow.xaml.cs(408, 38): error CS0246: The type or namespace name 'EchoStatus' could not be found (are you missing a using directive or an assembly reference?)

MainWindow.xaml.cs(392, 18): error CS0535: 'Host' does not implement interface member 'IHostApplication.Echo(string, MessageStatusEnum)'

MainWindow.xaml.cs(408, 58): error CS0103: The name 'EchoStatus' does not exist in the current context

MainWindow.xaml.cs(408, 49): error CS1750: A value of type '?' cannot be used as a default parameter because there are no standard conversions to type 'EchoStatus'

dmpas commented:
@EvilBeaver ээээ... а у тебя там рабочее пространство перед сборкой вычищается?.. Вот ну ведь не злодей же я, чтобы заливать то, что даже не компилируется :)

EvilBeaver commented:
Ошибка в xaml а в монодевелопе он вроде как выключен

dmpas commented:
@EvilBeaver едрить! ща пофикшу. ну или пофиксю.

EvilBeaver commented:
> ща пофикшу. ну или пофиксю

"Исправлю"?
